### PR TITLE
Added Command /shconfig toggle

### PR DESCRIPTION
**Pull request by Obsidianninja11**
Allows toggling boolean values in the config
Or by using value1 and value2, allows toggling between other value types. 

## Changelog Technical Details
+ Added Command /shconfig toggle <path> <value 1> <value 2>. - Obsidian

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt [hunk 1]
```diff
@@ -107,13 +108,42 @@ object SkyHanniConfigSearchResetCommand {
...
+
+        return try {
+            val (argsFilter) = createFilter(true) { path.lowercase() }
+            val (classFilter) = createFilter(false) {path.lowercase()}
+
+            val currentValue = findConfigElements(argsFilter, classFilter,true).toString()
+            val newValue = when (currentValue) {
+                 "[$rawJson1]" -> rawJson2
+                 "[$rawJson2]" -> rawJson1
+                else -> rawJson1
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]:
```suggestion
                 "[$rawJson1]" -> rawJson2
                else -> rawJson1
```

File: src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt [hunk 2]
```diff
@@ -107,13 +108,42 @@ object SkyHanniConfigSearchResetCommand {
...
+            }
+            setCommand(arrayOf("set", path, newValue))
+
+        } catch (e: Exception) {
+            ErrorManager.logError(e, "Error while trying to toggle config")
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]: 'logError(Throwable, String): Unit' is deprecated. Use data as well

File: src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt [hunk 3]
```diff
@@ -107,13 +108,42 @@ object SkyHanniConfigSearchResetCommand {
...
+    private suspend fun toggleCommand(args: Array<String>): String {
+        if (args.size < 2 || args.size == 3) return "§c/shconfig toggle <config name> [value 1] [value 2]"
+
+        val path = args[1]
+        val rawJson1 = if (args.size > 2)
+            if (args[2] == "clipboard") OSUtils.readFromClipboard() ?: return "§cClipboard has no string!" else args[2]
+        else "true"
+        val rawJson2 = if (args.size > 2)
+            if (args[3] == "clipboard") OSUtils.readFromClipboard() ?: return "§cClipboard has no string!" else args[3]
+        else "false"
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]: no clipboard for this

File: src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt [hunk 4]
```diff
@@ -107,13 +108,42 @@ object SkyHanniConfigSearchResetCommand {
...
-            "§eChanged config element $term."
+            "§eChanged config element $term to $rawJson."
         } catch (e: Exception) {
             ErrorManager.logError(e, "Could not change config element '$term' to '$rawJson'")
             "§cCould not change config element '$term' to '$rawJson'"
         }
     }
 
+    private suspend fun toggleCommand(args: Array<String>): String {
+        if (args.size < 2 || args.size == 3) return "§c/shconfig toggle <config name> [value 1] [value 2]"
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]:
```suggestion
        if (args.size == 1 || args.size == 3) return "§c/shconfig toggle <config name> [value 1] [value 2]"
```

File: src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt [hunk 5]
```diff
@@ -107,13 +108,42 @@ object SkyHanniConfigSearchResetCommand {
...
+        val rawJson1 = if (args.size > 2)
+            if (args[2] == "clipboard") OSUtils.readFromClipboard() ?: return "§cClipboard has no string!" else args[2]
+        else "true"
+        val rawJson2 = if (args.size > 2)
+            if (args[3] == "clipboard") OSUtils.readFromClipboard() ?: return "§cClipboard has no string!" else args[3]
+        else "false"
+
+        return try {
+            val (argsFilter) = createFilter(true) { path.lowercase() }
+            val (classFilter) = createFilter(false) {path.lowercase()}
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]: formatting

File: src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt [hunk 6]
```diff
@@ -107,13 +108,42 @@ object SkyHanniConfigSearchResetCommand {
...
+
+            val currentValue = findConfigElements(argsFilter, classFilter,true).toString()
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]: formatting

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
still dont get the point of toggling non-booleans, the case you described sounds like a bug in SkyHanni and this is a poor band-aid fix to it

File: src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt (hunk 3 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]: Why?

Comment (CalMWolfs) [reply src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]: Why is it useful to have clipboard here

Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]: Consistency?
Why is it bad to have clipboard here?

Comment (hannibal002) [reply src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]: I need to look at the full code later, but it looks like workarounds with clipboard logic. Im a fan of allowing to have `/shconfig set path clipboard` to work with json directly, same as outputting via `/shconfig load path`. Just Parse the clipboard content as string to a json element and try to apply in the config element. Same for reading

File: src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt [hunk 7]
```diff
@@ -107,13 +114,46 @@ object SkyHanniConfigSearchResetCommand {
...
+            }
+            setCommand(arrayOf("set", path, newValue))
+
+        } catch (e: Exception) {
+            ErrorManager.logErrorWithData(
+                e, "Error while trying to toggle config element",
+                "args" to args.toList(),
+                "path" to path,
+                "value1" to rawJson1,
+                "value2" to rawJson2,
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]: these three are already included in args.toList()

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
just this one change i think

File: src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt (hunk 7 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]: well value1 and value2 arent exactly but ig no point

File: src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt [hunk 8]
```diff
@@ -107,13 +114,43 @@ object SkyHanniConfigSearchResetCommand {
...
+        val path = args[1]
+        val rawJson1 = if (args.size > 2) args[2] else "true"
+        val rawJson2 = if (args.size > 2) args[3] else "false"
+
+        return try {
+            val (argsFilter) = createFilter(true) { path.lowercase() }
+            val (classFilter) = createFilter(false) { path.lowercase() }
+
+            val currentValue = findConfigElements(argsFilter, classFilter, true).toString()
+            val newValue = when (currentValue) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]: You can one line this with `val newValue = if (currentValue == "[$rawJson1]") rawJson2 else rawJson1`

File: src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt [hunk 9]
```diff
@@ -107,13 +114,43 @@ object SkyHanniConfigSearchResetCommand {
...
+
+        val path = args[1]
+        val rawJson1 = if (args.size > 2) args[2] else "true"
+        val rawJson2 = if (args.size > 2) args[3] else "false"
+
+        return try {
+            val (argsFilter) = createFilter(true) { path.lowercase() }
+            val (classFilter) = createFilter(false) { path.lowercase() }
+
+            val currentValue = findConfigElements(argsFilter, classFilter, true).toString()
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt]: please show the label for boolean parameters. e.g. `onlyValue = true`

File: src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt [hunk 10]
```diff
@@ -107,13 +114,43 @@ object SkyHanniConfigSearchResetCommand {
...
+            val newValue = when (currentValue) {
+                "[$rawJson1]" -> rawJson2
+                else -> rawJson1
+            }
+            setCommand(arrayOf("set", path, newValue))
+
+        } catch (e: Exception) {
+            ErrorManager.logErrorWithData(
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt:144]: Do we actually see this error message? An error would be cathced in the function setCommand and cant reach this function, right? can you test?

**Review by hannibal002 — CHANGES_REQUESTED** [review]
looks fine. have not tested in game

File: src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt (hunk 10 above)
Comment (Obsidianninja11) [reply src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt:144]: idk
I wasn't able to get one. so probably not

doesn't hurt to have though does it?

Comment (hannibal002) [reply src/main/java/at/hannibal2/skyhanni/test/SkyHanniConfigSearchResetCommand.kt:144]: Well its not needed, but also doesnt matter if we keep it. its fine.

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**